### ISSUE TITLE
It's possible to use resources in a `Context` where they were not created.

### ISSUE THREAD

**Issue by icefoxen** [ggez/ggez#326]
If you drop a `Context` and create a new one, you can feed resources from the old context into the new one, with undefined (but Safe, I think?) results.  This is basically a use-after-free problem.  At best, it panics as it tries to use a texture handle or something that no longer exists.  At worst, your `Image` or whatever will contain a texture handle that no longer points to what you think it does.

I don't *really* consider this a problem because I can't think of any case when you'd ever want to drop and re-create a `Context`.  Changing certain graphics settings perhaps?  You can never have more than one `Context` at the time because it will try to initialize SDL2 more than once and fail.

It's still worth pondering, and hedging if possible.  Thoughts:

 * Having lifetimes for every object pointing to the `Context` where it was created is something I consider unacceptably painful to use.
 * Having an `Rc` or something pointing to the `Context` in a resource is maybe acceptable, but a large change.  It's actually an interesting idea for `Context`-less API in general but that's a different story (see #298 and #295 ).
 * Having a timestamp or UUID or something in a `Context` and the resources created by it, and checking that they match whenever the resource is used, would work.  This is probably not something we want in the general case, but may be useful as a debugging tool via `debug_assert` checks or a feature flag or such.

Example, modification of `hello_world.rs`

```rust
extern crate ggez;
use ggez::conf;
use ggez::event;
use ggez::{Context, GameResult};
use ggez::graphics;
use std::env;
use std::path;

struct MainState {
    text: graphics::Text,
}

impl MainState {
    fn new(ctx: &mut Context, message: &str) -> GameResult<MainState> {
        let font = graphics::Font::new(ctx, "/DejaVuSerif.ttf", 48)?;
        let text = graphics::Text::new(ctx, message, &font)?;

        let s = MainState {
            text: text,
        };
        Ok(s)
    }
}

impl event::EventHandler for MainState {
    fn update(&mut self, _ctx: &mut Context) -> GameResult<()> {
        Ok(())
    }

    fn draw(&mut self, ctx: &mut Context) -> GameResult<()> {
        graphics::clear(ctx);

        let dest_point = graphics::Point2::new(10.0, 10.0);
        graphics::draw(ctx, &self.text, dest_point, 0.0)?;
        graphics::present(ctx);

        Ok(())
    }
}

pub fn main() {
    let state = &mut {
        let c = conf::Conf::new();
        let ctx = &mut Context::load_from_conf("helloworld", "ggez", c).unwrap();

        if let Ok(manifest_dir) = env::var("CARGO_MANIFEST_DIR") {
            let mut path = path::PathBuf::from(manifest_dir);
            path.push("resources");
            ctx.filesystem.mount(&path, true);
        }
        MainState::new(ctx, "Hello world!").unwrap()
    };

    let c = conf::Conf::new();
    let ctx2 = &mut Context::load_from_conf("helloworld", "ggez", c).unwrap();
    if let Ok(manifest_dir) = env::var("CARGO_MANIFEST_DIR") {
       let mut path = path::PathBuf::from(manifest_dir);
        path.push("resources");
        ctx2.filesystem.mount(&path, true);
    }
    let state2 = MainState::new(ctx2, "Goodbye world!").unwrap();


    // Here we use the first `MainState` we created but it displays "Goodbye world!",
    // using the texture handle from the second `MainState`.
    if let Err(e) = event::run(ctx2, state) {
        println!("Error encountered: {}", e);
    } else {
        println!("Game exited cleanly.");
    }
}
```

**Comment by icefoxen** [comment]
Servo has a cool way of doing something like the third option: https://github.com/servo/servo/pull/11442/files

**Comment by icefoxen** [comment]
Fixed in 23b6bca

Like smashing a fruit fly with a sledgehammer!
